Ransomware tutorial (#4181)
guardicore/monkey: docs/content/tutorials/ransomware.md

This tutorial assumes that you've completed [Tutorial 1: Hello, Monkey](../hello-monkey), since it builds off of the knowledge gained from that tutorial.

You'll learn:
- How to setup an environment for a ransomware scenario
- How to configure Infection Monkey to perform a ransomware attack
- How to observe the results of the ransomware attack

### Prerequisites
First, make sure that you have the following installed:
- `docker` and `docker-compose`

### Run the environment
Next, we'll use `docker compose` to run Infection Monkey along with our vulnerable container.

1. Download the following compose file: [docker-compose.yml](../hello-monkey/docker/docker-compose.yaml)

2. Then, navigate to the directory where you downloaded the file and run the following command to start the environment:

   ```
   docker compose up
   ```

Then, open a browser to [https://localhost:5000](https://localhost:5000), and login.

### Configure the vulnerable container
For this scenario, we're going to need some valuable data so that it can be held for ransom. We'll create a folder named `vault`, and we'll add a list of passwords to that vault.

{{% notice warning %}}

> Infection Monkey will encrypt the contents of whichever folder we direct it to target. Therefore, when setting up a ransomware scenario, make sure you have a way to recover that data, or target a folder with dummy data.

Reviewer: > whichever folder we direct it to target

This is perfectly correct language, but I feel like "direct" and "target" are _almost_ synonyms. If I didn't already know what you were talking about, this might throw me.

> Infection Monkey will encrypt the contents of whichever folder we direct it to target. Therefore, when setting up a ransomware scenario, make sure you have a way to recover that data, or target a folder with dummy data.

Reviewer: I'd recommend reversing order of suggestions:

```suggestion
Infection Monkey will encrypt the contents of whichever folder we direct it to target. Therefore, when setting up a ransomware scenario, target a folder with dummy data or make sure you have a way to recover the data in the target directory.
```

We'd prefer users use dummy data.

{{% /notice %}}

Connect to the container:

```
docker exec -it hello bash
```

Add a file named `passwords.txt`:
```

> su user

Reviewer: Can the "connect to the container" command be modified to connect you as a specific user? This would let the user skip 2 commands in the next step.

Author: Good suggestion. It looks like `docker exec` has a `--user` option so I'll use that

cd
mkdir vault
cat << EOF > vault/passwords.txt
password
P@ssw0rd
supersecretpassword
EOF
```

> Now let's see if we can get Infection Monkey to ransom our vault data!

Reviewer: ```suggestion
Now let's see if we can get Infection Monkey to encrypt our data!
```

### Configure Infection Monkey

#### Install plugins
Our first task is to make sure we have the required plugin installed. Of course we'll need to install the Ransomware plugin. We'll still need to exploit the container, so we'll install the SSH exploiter plugin as well.

> ![Plugins installed](../../images/tutorials/ransomware/1-plugins-installed.jpg)

Reviewer: Should this section be as descriptive as it is in the first tutorial? I don't want to repeat ourselves too much, but I think it's highly likely people will skip the 1st tutorial and go straight for the second one.

#### Tell the Monkey which exploiter to use
Make sure SSH Exploiter is enabled in the _Enabled exploiters_ list:

![Enable the SSH Exploiter](../../images/tutorials/hello-monkey/12-exploiter-enabled.jpg)

#### Tell the Monkey which machine to target

> Make sure `hello` is provided as a target under _Scan target list_:

Reviewer 2: I think it might be a good idea to explain what `hello` is somewhere in this tutorial too.

![Scan target list in the Network Analysis configuration](../../images/tutorials/hello-monkey/5-scan-target-list.jpg)

#### Tell the Monkey what credentials to use
Enter `user` for _Identity_ and `password` for _Password_:

![Credentials entered](../../images/tutorials/hello-monkey/13-credentials-input.jpg)

#### Tell the Monkey to use the Ransomware plugin
Next, we need to enable the Ransomware plugin. Select the **Payloads** tab, and enable the _Ransomware_ plugin in the _Enabled payloads_ list.

![Ransomware enabled](../../images/tutorials/ransomware/2-ransomware-enabled.jpg)

We also need to tell the Ransomware plugin which folder to hold for ransom. Select _Ransomware_ in the _Enabled payloads_ list in order to view its settings.

> Since the `hello` container is linux-based, and we want to target `/home/user/vault` folder, set the _Linux target directory_ to `/home/user/vault`.

Reviewer 2: ```suggestion
Since the `hello` container is linux-based, set the _Linux target directory_ to `/home/user/vault`.
```

Also make sure that _Leave ransom note_ is checked.

![Ransomware configured](../../images/tutorials/ransomware/3-ransomware-configuration.jpg)

Don't forget to **Submit** the configuration!

### Run Infection Monkey
Now that we've configured Infection Monkey, let's run it!

Go to the **Run Monkey** page and run the Monkey **From Island**.

> If you look at the **Infection Map** you'll see that the `hello` container gets exploited as it did in [Tutorial 1: Hello, Monkey](../hello-monkey).

Reviewer 2: ```suggestion
If you look at the **Infection Map**, you'll see that the `hello` container gets exploited as it did in [Tutorial 1: Hello, Monkey](../hello-monkey).
```

![Infection Monkey run completed](../../images/tutorials/hello-monkey/7-run-monkey.jpg)

Great, so the run has completed, but how do we know whether or not our ransomware succeeded?

One good place to look is the **Ransomware report**. Navigate to the **Security Reports** page, and select the **Ransomware report** tab. Then take a look at the **3. Attack** section. It lists the files that Infection Monkey was able to encrypt:

![List of files taken for ransom in ransomware report](../../images/tutorials/ransomware/4-ransomware-report.jpg)

It looks like our passwords are being held for ransom! Let's connect to the container and observe the contents of the vault:
```
docker exec -it hello bash

ls -1 /home/user/vault
```

We'll see two files:
```
README.txt
passwords.txt.m0nk3y
```

`README.txt` is the ransom note that Infection Monkey leaves in the directory targeted by the Ransomware plugin. `passwords.txt.m0nk3y` is our password list (formerly `passwords.txt`) that Infection Monkey encrypted. You can print the contents of the file to verify that the file is indeed encrypted:

```shell

> > cat /home/user/vault/passwords.txt.m0nk3y

Reviewer: I think it's also worth instructing the user to `cat README.txt`, though I don't know that we need to necessarily display the output in the tutorial. Although:
![image](https://github.com/guardicore/monkey/assets/19957806/13faee59-0ffb-48fe-8d04-5c3efaa1e5dc)

Reviewer 2: Maybe just link the file? They can crosscheck if they want to.

��������������ύ����������������������
```

Looks like the Monkey did its job!

{{% notice tip %}}

> Infection Monkey uses a bit flip algorithm to "encrypt" files. So if you find yourself in a situation where you need to reverse the encryption, you can drop the file extension that the monkey adds, and then re-run the ransomware and it will re-flip the bits, returning the files to their original state.

Reviewer 2: ```suggestion
Infection Monkey uses a bit flip algorithm to "encrypt" files. So if you find yourself in a situation where you need to reverse the encryption, you can drop the file extension that the Monkey adds, and then re-run the ransomware simulation. This will re-flip the bits, returning the files to their original state.
```

{{% /notice %}}

### Review
What have we learned?

> - Infection Monkey provides a Ransomware plugin which allows one to simulate a ransomware attack

Reviewer 2: ```suggestion
- Infection Monkey provides a Ransomware plugin that allows one to simulate a ransomware attack
```

> - How to specify a path for Infection Monkey to target for ransom in every machine that it exploits

Reviewer 2: "in every machine that it exploits"
Not entirely true. It'll run on every machine that runs an Agent.

```

`README.txt` is the ransom note that Infection Monkey leaves in the directory targeted by the Ransomware plugin. `passwords.txt.m0nk3y` is our password list (formerly `passwords.txt`) that Infection Monkey encrypted. First, let's take a peek at the ransom note:

```
$ cat ~/vault/README.txt
```

You should see [this](https://raw.githubusercontent.com/guardicore/monkey/develop/monkey/agent_plugins/payloads/ransomware/src/ransomware_readme.txt) printed to the console.

Next, let's print the contents of the encrypted password file to verify that the file is indeed encrypted:

```shell
$ cat ~/vault/passwords.txt.m0nk3y
��������������ύ����������������������
```

Looks like the Monkey did its job!

{{% notice tip %}}
Infection Monkey uses a bit flip algorithm to "encrypt" files. So if you find yourself in a situation where you need to reverse the encryption, you can drop the file extension that the Monkey adds, and then re-run the ransomware simulation. This will re-flip the bits, returning the files to their original state.
{{% /notice %}}

### Review
What have we learned?
- Infection Monkey provides a Ransomware plugin that allows one to simulate a ransomware attack

> - How to specify a path for Infection Monkey to target for ransom in every machine to whom it successfully propagates

Reviewer 2: ```suggestion
- How to specify a path for Infection Monkey's ransomware simulation to target in every machine on which an Agent runs
```

to run the ransomware simulation on the exploited container, we'll need the
ransomware plugin.

Navigate to the **Plugins** page by selecting _Plugins_ in the navigation
sidebar. You'll see a list of all the plugins that can be installed. Install
the _SSH Exploiter_ and _Ransomware Payload_ by clicking their respective
download icons in the right-most column of the table. You'll see the download
icons transform into check marks once installation is complete.

![Plugins page](../../images/tutorials/ransomware/010-plugin-installation.gif)

Now that the plugins have been installed, we're ready to configure our
simulation.

#### Tell the Monkey which exploiter to use
Navigate to the **Configuration** page by selecting _Configure Monkey_ on the
**Getting Started** page (or select _Configuration_ in the navigation sidebar).
Check the box next to the _SSH Exploiter_ to enable it.

![Enable the SSH
Exploiter](../../images/tutorials/ransomware/020-exploiter-enabled.jpg)

#### Tell the Monkey which machine to target
Switch to the **Network analysis** subtab and click the yellow _+_ button under
_Scan target list_. You'll see a new field appear. Enter the hostname of the

> vulnerable container, `hello` in the field.

Author: ```suggestion
vulnerable container, `hello`, in the field.
```

(formerly `passwords.txt`) that Infection Monkey encrypted. First, let's take a
peek at the ransom note:

```
$ cat ~/vault/README.txt
```

You should see
[this](https://raw.githubusercontent.com/guardicore/monkey/develop/monkey/agent_plugins/payloads/ransomware/src/ransomware_readme.txt)
printed to the console.

Next, let's print the contents of the encrypted password file to verify that
the file is indeed encrypted:

```shell
$ cat ~/vault/passwords.txt.m0nk3y
��������������ύ����������������������
```

Looks like the Monkey did its job!

{{% notice tip %}}
Infection Monkey uses a bit flip algorithm to "encrypt" files. So if you find
yourself in a situation where you need to reverse the encryption, you can drop
the file extension that the Monkey adds, and then re-run the ransomware
simulation. This will re-flip the bits, returning the files to their original
state.
{{% /notice %}}

### Review
In this tutorial, you have learned:
- Infection Monkey provides a ransomware plugin that enables you to simulate a
  ransomware attack.
- How to prepare systems for a ransomware simulation by deploying target files.

> - How to configure Infection Monkey to use it's ransomware plugin.

Author: ```suggestion
- How to configure Infection Monkey to use its ransomware plugin.
```
